push_test/pushclient.py

```diff
@@ -0,0 +1,335 @@
+                        version=version or notif['version'])
+                await self._send(messageType="ack",
+                                 channelID=channelID or notif['channelID'],
+                                 version=version or notif['version'],
+                                 no_recv=True)
+        except IndexError:
+            pass
+        cmd, args = self._next_task()
+        await getattr(self, cmd)(**args)
+
+    async def cmd_register(self, channelID=None, key=None, **kwargs):
+        """Register a new ChannelID
+
+        :param channelID: UUID for the channel to register
+        :param key: applicationServerKey for a restricted access channel
+        :param kwargs: additional optional arguments
+        :return:
+
+        """
+        log_msg(out=self.output,
+                status="Sending new channel registration")
+        channelID = channelID or self.channelID or str(uuid.uuid4())
+        args = dict(messageType='register',
+                    channelID=channelID)
+        if key:
+            args[key] = key
+        args.update(kwargs)
+        await self._send(**args)
+
+    async def cmd_done(self, **kwargs):
+        """Close all connections and mark as done
+
+        :param kwargs: ignored
+        :return:
+
+        """
+        log_msg(out=self.output,
+                status="done")
+        await self.cmd_close()
+        await self.connection.close_connection()
+
+    async def recv_hello(self, **msg):
+        """Process a received "hello"
+
+        :param msg: body of response
+        :return:
+
+        """
+        assert(msg['status'] == 200)
+        try:
+            self.uaid = msg['uaid']
+            cmd, args = self._next_task()
+            await getattr(self, cmd)(**args)
+        except KeyError as ex:
+            raise PushException from ex
+
+    async def recv_register(self, **msg):
+        """Process a received registration message
+
+        :param msg: body of response
+        :return:
+
+        """
+        assert(msg['status'] == 200)
+        self.pushEndpoint = msg['pushEndpoint']
+        self.channelID = msg['channelID']
+        log_msg(out=self.output,
+                flow="input",
+                msg=dict(
+                message="register",
+                channelID=self.channelID,
+                pushEndpoint=self.pushEndpoint))
+        cmd, args = self._next_task()
+        await getattr(self, cmd)(**args)
+
+    async def recv_notification(self, **msg):
+        """Process a received notification message
+
+        :param msg: body of response
+
+        """
+        def repad(str):
+            return str + '===='[len(msg['data']) % 4:]
+
+        msg['_decoded_data'] = base64.urlsafe_b64decode(
+            repad(msg['data'])).decode()
+        log_msg(out=self.output,
+                flow="input",
+                message="notification",
+                msg=msg)
+        self.notifications.append(msg)
```

Comment:
These command functions would be simpler if they didn't all mutate object state. How come they can't act more like functions and return the msg so that the caller can decide what to do?